sklearn/manifold/tests/test_t_sne.py

```diff
@@ -921,6 +921,27 @@ def test_tsne_with_legacy_euclidean_squaring(method):
     assert_array_equal(X_transformed_tsne, X_transformed_tsne_precomputed)
 
 
+@pytest.mark.parametrize('method', ['exact', 'barnes_hut'])
+@ignore_warnings(category=FutureWarning)
+def test_tsne_with_different_squaring_methods(method):
+    """Isolate failing behavior in
+    `test_tsne_with_legacy_euclidean_squaring.`"""
+    random_state = check_random_state(0)
+    n_components_original = 3
+    n_components_embedding = 2
+    X = random_state.randn(50, n_components_original).astype(np.float32)
+    X_precomputed_1 = pairwise_distances(X, metric='euclidean') ** 2
+    X_precomputed_2 = pairwise_distances(X, metric='euclidean', squared=True)
+    X_transformed_tsne_precomputed_1 = TSNE(
+        metric='precomputed', n_components=n_components_embedding,
+        random_state=0, method=method).fit_transform(X_precomputed_1)
+    X_transformed_tsne_precomputed_2 = TSNE(
+        metric='precomputed', n_components=n_components_embedding,
+        random_state=0, method=method).fit_transform(X_precomputed_2)
+    assert_array_equal(X_transformed_tsne_precomputed_1,
+                       X_transformed_tsne_precomputed_2)
+
+
```

Comment:
While making the requested change in 08d7b79793f795eb8c370837563f72a899e9ce30, I discovered a separate issue: using different squaring methods results in different TSNE results, i.e. the test here fails.

The reason the difference is relevant is because of `'barnes_hut'` method. 
* If distances are not precomputed, it uses `**=2` to square [here](https://github.com/scikit-learn/scikit-learn/blob/5edf203d6ee8a168c127b7bff7410ac83b497a30/sklearn/manifold/_t_sne.py#L809). 
* If distances are precomputed, `squared=True` could be used, e.g. in [this other test](https://github.com/scikit-learn/scikit-learn/pull/17662/files#diff-bf12bea42d63e12299a1eee54411bcc6R916).

These two cases produce different results. I believe this behavior was present before any of the changes in this PR.

Reply:
To be clear: [`test_tsne_with_legacy_euclidean_squaring`](https://github.com/scikit-learn/scikit-learn/pull/17662/files#diff-bf12bea42d63e12299a1eee54411bcc6R902) currently passes when `exact` is used, and fails when `barnes_hut` is used. The above behavior is why.

Reply:
I explained this in an unclear way. I'm going to try and reword my thoughts.

Reply:
```
X_precomputed_1 = pairwise_distances(X, metric='euclidean') ** 2
X_precomputed_2 = pairwise_distances(X, metric='euclidean', squared=True)
```

Although mathematically the same, they might slightly differ in floating arithmetics. Then we can expect the results of t-sne to be slightly different too. Usually we use `assert_allclose` to compare arrays of floats. Since you're using float32, you can change the rtol to 1e-6 or even 1e-5.

Reply:
Ah, sorry, it might have been better to also include the output of the failing tests.

```
E       Mismatched elements: 100 / 100 (100%)
E       Max absolute difference: 182.95767
E       Max relative difference: 35.374535
E        x: array([[-36.36463 , -60.091194],
E              [ 35.217983, -77.36719 ],
E              [-23.366423,  -1.641921],...
E        y: array([[ -83.61592 ,  -40.051804],
E              [ -12.799014,  105.59048 ],
E              [  -2.169685,  -25.965828],...
```

A small change in the distances (e.g. due to float arithmetic) seems to have caused a large difference in the TSNE output. So, I'm not sure `assert_allclose` would help here, but I could be mistaken.

Reply:
I'm not entirely surprised that tSNE is unstable. I can confirm that the differences are small:

absolute diff between `X_precomputed_1` and `X_precomputed_2`:
![absdiff](https://user-images.githubusercontent.com/78827/85648625-7cc05680-b6e4-11ea-9e51-7b4e48f36393.png)

Relative diff:
![reldiff](https://user-images.githubusercontent.com/78827/85648632-81850a80-b6e4-11ea-9cae-bca5b66af2b9.png)

> Although mathematically the same, they might slightly differ in floating arithmetics.

More explicitly, in `X_precomputed_1 = pairwise_distances(X, metric='euclidean') ** 2` the computation loses precision when taking the square root in euclidean_distances.

Reply:
> I'm not entirely surprised that tSNE is unstable. I can confirm that the differences are small:

Thank you for taking the time to verify this, and to share the plots you've created. :)

Reply:
Are there perhaps parameters we can set to make the tSNE optimisation more stable? I would have still expected to land up with the same nearest neighbors graph, but slightly different probability distribution.

Reply:
Quick question: Would it be better to have this chat in a separate issue, just to open up the conversation a bit? 

While distance squaring was what started the conversation, from my understanding any small differences could trigger this behaviour, so it might be a bit off-topic from this PR.

Reply:
I'm not sure if the instability of tSNE is an issue as such... But it's an issue with regards to how to test this change!

Reply:
Aha, that's good to know! I had assumed instability was a general issue, so thank you for clarifying. :)

A quick/dirty way to fix the `test_tsne_with_legacy_euclidean_squaring` test would be to replace [this line](https://github.com/scikit-learn/scikit-learn/pull/17662/files#diff-bf12bea42d63e12299a1eee54411bcc6R916) with:

```python
if method == 'barnes_hut':
    X_precomputed = pairwise_distances(X, metric='euclidean') ** 2
else:
    X_precomputed = pairwise_distances(X, metric='euclidean', squared=True)
```
That way the squaring method matches up with what's used internally in `TSNE`. But, if I understand correctly, you're asking if there's a way to not have to resort to this?

Reply:
Going out on a limb here, but the lack of stability could also have something do with with the use of `randn` to generate sample pairwise distances. Without much structure in our sample data, t-SNE might not find meaningful topology in the data, as is shown in [this tutorial](https://scikit-learn.org/stable/auto_examples/manifold/plot_t_sne_perplexity.html#sphx-glr-auto-examples-manifold-plot-t-sne-perplexity-py).

I wonder if it would be better to create/reuse a small toy dataset for TSNE tests that has structure. This could better reflect the intended use cases for t-SNE.

Reply:
Yep, the test pass with `make_blobs`. I''ll update accordingly.

Reply:
Yes, it does seem like `make_blobs` fixed the stability issues. :)

Reply:
That makes a lot of sense. Good idea :)

Reply:
Thank you for your help! Your questions pointed me in the right direction. :)